third_party/move/move-model/src/builder/model_builder.rs

```diff
@@ -460,6 +457,161 @@ impl<'env> ModelBuilder<'env> {
+        ));
+        loop_notes
+    }
+
+    /// Checks if a field type occurs in its access path,
+    /// and adds diagnostics for recursive struct definitions if `ty` occurs in `path` and `ty` is also the struct type `checking`.
+    ///
+    /// Returns true iff no resursive definition for `checking` is found at this level.
+    ///
+    /// `path`: `path[0]` contains `path[1]` ... `path[-1]` contains `ty`
+    /// `checking`: the root struct which we are checking for recursive definition for
+    /// `loc_checking`: location of the struct we are checking for
+    fn check_recusive_struct_with_parents(
+        &self,
+        ty: &Type,
+        ty_loc: Loc,
+        path: &mut Vec<(Loc, Symbol, Type)>,
+        checking: QualifiedId<StructId>,
+        loc_checking: &Loc,
+    ) -> bool {
+        match ty {
+            Type::Struct(mid, sid, insts) => {
+                let this_struct_id = mid.qualified(*sid);
+                let this_struct_entry = self.lookup_struct_entry(this_struct_id);
+                // checks if `ty` occurs in `path`
+                for (_parent_loc, _field_symbol, parent) in path.iter() {
+                    if let Type::Struct(mid, sid, _) = parent {
+                        let parent_id = mid.qualified(*sid);
+                        if parent_id == this_struct_id {
+                            if checking == this_struct_id {
+                                let loop_notes = self.gen_error_msg_for_fields_loop(
+                                    path,
+                                    this_struct_id,
+                                    ty_loc.clone(),
+                                );
+                                self.error_with_notes(
+                                    loc_checking,
+                                    &format!(
+                                        "recursive definition {}",
+                                        self.get_struct_display_name_simple(this_struct_id)
+                                    ),
+                                    loop_notes,
+                                );
+                                return true;
+                            } else {
+                                return false;
+                            }
+                        }
+                    } else {
+                        panic!("field access path contains non-struct")
+                    }
+                }
+                if let Some(fields) = &this_struct_entry.fields {
+                    // check for descendant fields recursively
+                    for (field_name, (field_loc, _field_idx, field_ty_uninstantiated)) in
+                        fields.iter()
+                    {
+                        path.push((ty_loc.clone(), *field_name, ty.clone()));
+                        let field_ty_instantiated = field_ty_uninstantiated.instantiate(insts);
+                        // short-curcuit upon first recursive occurence found
```

Comment:
```suggestion
                        // short-circuit upon first recursive occurrence found
```

Reply:
Also, we are not doing what this comment claims. It is because the return values are flipped.

Please add this test case (or similar), which demonstrates the issue:

```
struct X {
    f: Y,
    g: Y,
}

struct Y {
    f: X,
    g: X
}
```

Reply:
done